src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,334 @@
+    }
+    return itps;
+  }
+
+  // Experimental!
+  @Override
+  public List<BooleanFormula> getTreeInterpolants(
+      List<? extends Collection<Term>> pPartitionedFormulas, int[] pStartOfSubTree)
+      throws SolverException, InterruptedException {
+    Preconditions.checkState(!closed);
+    assert InterpolatingProverEnvironment.checkTreeStructure(
+        pPartitionedFormulas.size(), pStartOfSubTree);
+
+    // Generate every Interpolation Pair
+    ArrayList<ArrayList<ArrayList<Collection<Term>>>> interpolationPairs =
+        getTreeInterpolationPairs(pPartitionedFormulas, pStartOfSubTree);
+
+    final ArrayList<Term> itps = new ArrayList<>();
+    try { // Interpolate every Interpolation Pair
+      for (int i = 0; i < interpolationPairs.size(); i++) {
+        itps.add(getCVC5Interpolation(interpolationPairs.get(i)));
+      }
+    } catch (UnsupportedOperationException e) {
+      if (e.getMessage() != null) {
+        throw new SolverException(e.getMessage(), e);
+      } else {
+        throw e;
+      }
+    }
+
+    final List<BooleanFormula> result = new ArrayList<>();
+    for (Term itp : itps) {
+      result.add(creator.encapsulateBoolean(itp));
+    }
+    assert result.size() == pStartOfSubTree.length - 1;
+    return result;
+  }
+
+  /**
+   * Interpolates a Tuple of Interpolants according to Craig-Interpolation using CVC5-Interpolation.
+   *
+   * @param formulaPair Interpolation Pair
+   * @return Interpolation of the Interpolation Pair following the definition of
+   *     Craig-Interpolation.
+   */
+  private Term getCVC5Interpolation(ArrayList<ArrayList<Collection<Term>>> formulaPair) {
+    assert formulaPair.size() == 2; // Check that formulaPair is a Tuple
+
+    ArrayList<Collection<Term>> assertedInterpols = formulaPair.get(0);
+    ArrayList<Collection<Term>> addedInterpols = formulaPair.get(1);
+
+    // Respect Asserted Formulas not in the Interpolation Pairs
+    ArrayList<Collection<Term>> combinedInterpols = new ArrayList<>();
+    combinedInterpols.addAll(assertedInterpols);
+    combinedInterpols.addAll(addedInterpols);
+
+    Collection<Term> extraAssertions = getAssertedTermsNotInCollection(combinedInterpols);
+    Term extraAssert = new Term();
+
+    if (extraAssertions.isEmpty()) {
+      extraAssert = solver.mkTrue();
+    } else {
+      extraAssert = buildConjunctionOfFormulas(extraAssertions);
+    }
+
+    Term phim = buildConjunctionOfFormulasOverArray(assertedInterpols);
+    Term phip = buildConjunctionOfFormulasOverArray(addedInterpols);
+
+    solver.resetAssertions();
+    solver.assertFormula(solver.mkTerm(Kind.AND, extraAssert, phim));
+
+    // with Phip negated, CVC5 Interpolation produces an equivalent interpolation to
+    // Craig-Interpolation
+    Term interpolant = solver.getInterpolant(solver.mkTerm(Kind.NOT, phip));
+
+    // restore Assertion state
+    solver.resetAssertions();
+
+    assertedFormulaHash.forEach((n) -> solver.assertFormula(n));
```

Comment:
This should rebuild the solver-context, maybe. 
1) Why is the code between deleting the solver-context and rebuilding it not in a try-finally-block?
2) The rebuild ignores levels from push/pop.

Reply:
With the separate solver-instance, rebuilding is not needed anymore.